source/core/Variants.cpp

```diff
@@ -19,25 +19,117 @@ SDG
 #include <map>
 #include "Variants.h"
 
+namespace
+{
+
+using namespace Vireo;
+bool IsNullVariant(VariantTypeRef variant)
+{
+    return variant == nullptr;
+}
+
+VariantTypeRef CreateNewVariantFromType(TypeRef inputType)
+{
+    TypeManagerRef tm = THREAD_TADM();
+    VariantTypeRef newVariant = VariantType::New(tm, inputType == nullptr ? tm->BadType() : inputType);
+    return newVariant;
+}
+
+VariantTypeRef CreateNewVariantFromStaticTypeAndData(const StaticTypeAndData& input)
+{
+    TypeRef inputType = input._paramType;
+    TypeManagerRef tm = THREAD_TADM();
+
+    auto inputData = input._pData;
+    if (inputType->IsVariant()) {
+        VariantTypeRef variant = *reinterpret_cast<VariantTypeRef *>(input._pData);
+        inputType = variant->_underlyingTypeRef;
+        inputData = variant->Begin(kPARead);
+    }
+    VariantTypeRef newVariant = VariantType::New(tm, inputType);
+    newVariant->CopyData(inputData, newVariant->Begin(kPAWrite));
+    return newVariant;
+}
+
+VariantTypeRef CreateNewVariantFromVariant(const VariantTypeRef& inputVariant)
+{
+    TypeManagerRef tm = THREAD_TADM();
+    TypeRef underlyingType = inputVariant->_underlyingTypeRef;
+    VariantTypeRef newVariant = VariantType::New(tm, underlyingType);
+    newVariant->CopyData(inputVariant->Begin(kPARead), newVariant->Begin(kPAWrite));
+    return newVariant;
+}
+
+void SetVariantToDataTypeError(TypeRef inputType, TypeRef destType, ErrorCluster* errPtr)
+{
+    if (inputType && inputType->IsCluster() && destType->IsArray()) {
+        errPtr->SetErrorAndAppendCallChain(true, kUnsupportedOnTarget, "Variant To Data");
+    } else {
+        errPtr->SetErrorAndAppendCallChain(true, kVariantIncompatibleType, "Variant To Data");
+    }
+}
+
+}
+
 namespace Vireo
 {
+
+VariantType* VariantType::New(TypeManagerRef typeManager, TypeRef type)
+{
+    return TADM_NEW_PLACEMENT(VariantType)(typeManager, type);
+}
+
+//------------------------------------------------------------
+VariantType::VariantType(TypeManagerRef typeManager, TypeRef type)
+    : WrappedType(typeManager, type)
+{
+    _isFlat = false;
+    _ownsDefDefData = true;
+    _isMutableValue = true;
+    _hasCustomDefault = true;
+    _topAQSize = sizeof(void*);
+    _aqAlignment = sizeof(void*);
+    TypeRef copyInputType = DefaultValueType::New(typeManager, type, true);
+    _underlyingTypeRef = copyInputType;
+    _attributeMap = nullptr;
+    _wrapped = copyInputType;
+}
+
+NIError VariantType::ClearData(void*)
```

Comment:
I think we might need to implement InitData and CopyData and the overloads that a count parameter. This might be required for Array support to work.